### PR TITLE
Charts and tables - download dataset

### PR DESCRIPTION
**Pull request by rob-duplock**
Download dataset button now available for charts.

### CODE REVIEW

File: src/Plugin/Visualisation/Style/AxisChart.php [hunk 1]
```diff
@@ -836,4 +857,37 @@ protected function setArrayOrder(array $array_to_order, $weight_key = 'weight')
...
+  /**
+   * Returns the URI of a DVF file or dataset (E.g. for downloading).
+   *
+   * @param \Drupal\Core\Entity\EntityInterface $entity
+   *   The Drupal entity that contains the dvf_url or dvf_file.
+   *
+   * @return string
+   *   The URI of the file or dataset.
+   */
+  protected function getDatasetDownloadUri($entity) {
```
Comment (jez500) [inline src/Plugin/Visualisation/Style/AxisChart.php]: @rob-duplock  Does this account for future possible uris where dataset might be a JSON endpiont, view, or dataset that does not have a download link?

Maybe an additional method should be added to identify if something is downloadable and that determines button visibility and is also guardcode for a method like this one

File: src/Plugin/Visualisation/Style/AxisChart.php [hunk 2]
```diff
@@ -685,6 +685,20 @@ public function build() {
         '#type' => 'container',
         '#attributes' => ['class' => ['table-chart--actions']],
       ];
+
+      $file_uri = $this->getDatasetDownloadUri($this->visualisation->getEntity());
```
Comment (jez500) [inline src/Plugin/Visualisation/Style/AxisChart.php:690]: Noting that is only on an Axis chart, what about a visualisation that is not an axis chart, eg table or future vis.

feels like `getDatasetDownloadUri` should be abstracted to parent class and not specifically on axis only

**Review by jez500 — COMMENTED** [review]
Hey @rob-duplock looks good in general but a few comments regarding the approach before I would be happy with approval

**Comment by rob-duplock** [comment]
Thanks, @jez500 this is ready for another look.

**Review by jez500 — APPROVED** [review]

**Pull request merged**